File: backend/src/main/java/com/example/backend/service/MessageService.java
package com.example.backend.service;

import java.util.List;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import com.example.backend.model.entity.Conversation;
import com.example.backend.model.entity.Message;

public interface MessageService {
    // tạo hoặc lấy cuộc hội thoại 1-1
    Conversation createOrGetPrivateConversation(Long userId1, Long userId2);

    // tạo nhóm chat
    Conversation createGroupConversation(Long creatorId, String name, List<Long> memberIds);

    // gửi tin nhắn
    Message sendMessage(Long senderId, Long conversationId, String content, String mediaUrl, String emoji);

    // thu hồi tin nhắn
    Boolean recallMessage(Long senderId, Long messageId);

    // lấy danh sách tin nhắn
    Page<Message> getMessages(Long userId, Long conversationId, Pageable pageable);

    // đánh dấu đã xem
    Boolean markConversationAsSeen(Long userId, Long conversationId);
}

File: frontend/src/service/impl/MessageService.ts
import { http } from "../http/http";
import type { Conversation, Message } from "../../utils/EntityType";
import type { ApiDataResponse, Page } from "../../utils/ResponType";

export const createOrGetPrivateConversation = (currentUserId: number, userId2: number) => {
    return http.post<ApiDataResponse<Conversation>>(
        `/messages/conversations/private/${userId2}?currentUserId=${currentUserId}`
    );
};

export const createGroupConversation = (currentUserId: number, name: string, memberIds: number[]) => {
    return http.post<ApiDataResponse<Conversation>>(
        `/messages/conversations/group?currentUserId=${currentUserId}&name=${encodeURIComponent(name)}`,
        memberIds
    );
};

export const sendMessage = (
    currentUserId: number,
    conversationId: number,
    content?: string,
    mediaUrl?: string,
    emoji?: string
) => {
    const params = new URLSearchParams();
    params.append("currentUserId", currentUserId.toString());
    params.append("conversationId", conversationId.toString());
    if (content !== undefined) params.append("content", content);
    if (mediaUrl !== undefined) params.append("mediaUrl", mediaUrl);
    if (emoji !== undefined) params.append("emoji", emoji);

    return http.post<ApiDataResponse<Message>>(`/messages?${params.toString()}`);
};

export const recallMessage = (currentUserId: number, messageId: number) => {
    return http.delete<ApiDataResponse<boolean>>(`/messages/${messageId}/recall?currentUserId=${currentUserId}`);
};

export const getMessages = (currentUserId: number, conversationId: number, page: number = 0, size: number = 20) => {
    return http.get<ApiDataResponse<Page<Message>>>(
        `/messages/conversations/${conversationId}?currentUserId=${currentUserId}&page=${page}&size=${size}`
    );
};

export const markConversationAsSeen = (currentUserId: number, conversationId: number) => {
    return http.put<ApiDataResponse<boolean>>(`/messages/conversations/${conversationId}/seen?currentUserId=${currentUserId}`);
};
